airbase/download_api.py

```diff
@@ -0,0 +1,319 @@
+
+import asyncio
+from collections import defaultdict
+from enum import IntEnum
+from pathlib import Path
+from types import SimpleNamespace
+from typing import NamedTuple
+from urllib import parse
+from warnings import warn
+
+import aiofiles
+import httpx
+from tqdm import tqdm
+
+API_SERVICE_ROOT = "https://eeadmz1-downloads-api-appservice.azurewebsites.net"
+COUNTRY_CODES = set(
+    """
+    AD AL AT BA BE BG CH CY CZ DE DK EE ES FI FR GB GI GR HR HU
+    IE IS IT LI LT LU LV ME MK MT NL NO PL PT RO RS SE SI SK TR
+    XK
+    """.split()
+)
+
+DEFAULT = SimpleNamespace(
+    encoding="UTF-8",
+    progress=False,
+    raise_for_status=True,
+    max_concurrent=10,
+)
+
+
+class Dataset(IntEnum):
+    """
+    1. Unverified data transmitted continuously (Up-To-Date/UTD/E2a) data from the
+    beginning of 2023.
+    2. Verified data (E1a) from 2013 to 2022 reported by countries by 30 September each
+    year for the previous year.
+    3. Historical Airbase data delivered between 2002 and 2012 before Air Quality
+    Directive 2008/50/EC entered into force.
+
+    https://eeadmz1-downloads-webapp.azurewebsites.net/content/documentation/How_To_Downloads.pdf
+    """
+
+    Historical = Airbase = 3
+    Verified = E1a = 2
+    Unverified = UDT = E2a = 1
+
+    def __str__(self) -> str:
+        return self.name
+
+
+class DownloadInfo(NamedTuple):
+    """
+    info needed for requesting the URLs for one pollutant from one country and dataset
+    the request can be further restricted with the `cities` param
+    """
+
+    pollutant: str
+    country: str
+    dataset: Dataset
+    cities: tuple[str, ...] | None = None
+    source: str = "API"  # for EEA internal use
+
+    def request_info(self) -> dict[str, list[str] | list[Dataset] | str]:
+        return dict(
+            countries=[self.country],
+            cities=[] if self.cities is None else list(self.cities),
+            properties=[self.pollutant],
+            datasets=[self.dataset],
+            source=self.source,
+        )
+
+    @classmethod
+    def historical(
+        cls, pollutant: str, country: str, *cities: str
+    ) -> DownloadInfo:
+        return cls(pollutant, country, Dataset.Historical, cities)
+
+    @classmethod
+    def verified(
+        cls, pollutant: str, country: str, *cities: str
+    ) -> DownloadInfo:
+        return cls(pollutant, country, Dataset.Verified, cities)
+
+    @classmethod
+    def unverified(
+        cls, pollutant: str, country: str, *cities: str
+    ) -> DownloadInfo:
+        return cls(pollutant, country, Dataset.Unverified, cities)
+
+
+async def _raise_for_status(response: httpx.Response):
+    response.raise_for_status()
+
+
+async def _warn_for_status(response: httpx.Response):
+    if response.status_code >= 400:
+        warn(
+            f"HTTP Error {response.status_code} for URL {response.url}",
+            category=RuntimeWarning,
+        )
+
+
+class DownloadAPIClient(httpx.AsyncClient):
```

Comment:
In general I prefer module/functions + dependency injection  over class/methods + inheritance.

Here there is one potential advantage, we could reuse the client.
This could be done if the class is a contest manager.

I'll modify my PR to show what I mean.

Reply:
The `httpx.AsyncClient` actually is a context manager: https://www.python-httpx.org/advanced/clients/#usage

The only thing is that exiting the context closes the connections, when we actually want to re-use them.

Reply:
Now I see that you are using a different library 🤦🏼‍♂️ 
I updated my PR to show what I mean about `DownloadClient` as a context manager.
The tests show how it can be used. You should get a good idea of what I mean once add CLI sub commands for the new API.

Reply:
I updated my PR with separate CLI sub command for each dataset (`historical`/`verified`/`unverified`).

Here is the `historical` CLI sub command

https://github.com/JohnPaton/airbase/blob/9ea21fe9b424a57400160ff0444fc8ff13554017/airbase/cli.py#L136-L157

And the shared code for `historical`/`verified`/`unverified`

https://github.com/JohnPaton/airbase/blob/9ea21fe9b424a57400160ff0444fc8ff13554017/airbase/cli.py#L122-L134

maybe `DownloadClient` should have specialized download functions for `historical`/`verified`/`unverified` data?

Reply:
Haha, yeah sorry I should've mentioned that. I switched to `httpx` because `aiohttp` only supports context manager mode, but not having a long-lived client that can be called multiple times. 

I'll check out your changes. I agree with having specific high-level and client methods for the different datasets. Also from the user perspective it probably makes sense to treat them separately, and that's what EEA seems to be doing as well.